emu/src/cpu/data_processing.rs

```diff
@@ -24,10 +25,55 @@ enum PsrOpKind {
     MsrFlg,
 }
 
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum ShiftOperator {
+    Immediate(u32),
+    Register(u32),
+}
+
+impl std::fmt::Display for ShiftOperator {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Immediate(value) => write!(f, "#{value}"),
+            Self::Register(register) => write!(f, "{register}"),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum AluSecondOperandInfo {
+    Register {
+        op_kind: ShiftOperator,
+        shift_kind: ShiftKind,
+        register: u32,
+    },
+    Immediate {
+        base: u32,
+        shift: u32,
+    },
+}
+
+impl std::fmt::Display for AluSecondOperandInfo {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match *self {
+            Self::Register {
+                op_kind,
+                shift_kind,
+                register,
+            } => {
+                write!(f, "{register}, {shift_kind} {op_kind}")
```

Comment:
Here we miss a bit of logic in the case of immediate-specified shift, if we do a `LSL #0` it means that we don't do any shift so we don't show anything. `LSR #0` and `ASR #0` they mean `LSR #32`, `ASR #32`.
`ROR #0` means `RRX`

Reply:
Something like this:

```rust
 let amount = operand2.get_bits(7..=11);

                        if shift_type == ShiftKind::Ror && amount == 0 {
                            format!("R{rm}, RRX")
                        } else if shift_type == ShiftKind::Lsl && amount == 0 {
                            format!("R{rm}")
                        } else if amount == 0 {
                            format!("R{rm}, {shift_type} #32")
                        } else {
                            format!("R{rm}, {shift_type} #{amount}")
                        }
```